File: chalenge/jolly_jumpers/jolly.erl
-module (jolly).
-include_lib("eunit/include/eunit.hrl").

% from trip.erl
subtract(A, B) -> subtract(A, B, []).
subtract([A|H], [B|T], L) -> subtract(H, T, L ++ [abs(A-B)]);
subtract([], [], L) -> L.

differences(A) ->
  List1 = lists:sublist(A, length(A)-1),
  List2 = lists:sublist(A, 2, length(A)),
  subtract(List1, List2).
  
is_jolly(L) ->
  lists:sort(differences(L)) == lists:seq(1, length(L)-1).
  
  
differences_test_() -> [
  ?_assert([3,2,1] == differences([1,4,2,3]))
].

minimum_transfer_test_() -> [
  ?_assert(true == is_jolly([1, 4, 2, 3])),
  ?_assert(false == is_jolly([1, 4, 2, -1, 6]))
].

File: chalenge/jolly_jumpers/jolly.rb
class Array
  def jolly?
    differences.sort == (1...length).to_a
  end
  
  def differences
    self[0..-2].zip(self[1..-1]).map{|x| (x[0] - x[1]).abs }
  end
end

describe "Jolly Jumpers" do
  it "서로 인접해 있는 두 수의 차를 계산한다" do
    [1,4,2,3].differences.should == [3,2,1]
  end
  
  it "유쾌한 점퍼인가?" do
    [1, 4, 2, 3].should be_jolly
    [1, 4, 2, -1, 6].should_not be_jolly
  end
end